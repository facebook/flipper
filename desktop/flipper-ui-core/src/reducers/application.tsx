/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {v1 as uuidv1} from 'uuid';
import {getRenderHostInstance} from 'flipper-frontend-core';
import {Actions} from './';

export type LauncherMsg = {
  message: string;
  severity: 'warning' | 'error';
};

export type StatusMessageType = {
  msg: string;
  sender: string;
};

type SubShareType =
  | {
      type: 'file';
      file: string;
    }
  | {
      type: 'link';
      url?: string;
    };

export type ShareType = {
  statusComponent?: React.ReactNode;
  closeOnFinish: boolean;
} & SubShareType;

export type State = {
  leftSidebarVisible: boolean;
  rightSidebarVisible: boolean;
  rightSidebarAvailable: boolean;
  windowIsFocused: boolean;
  share: ShareType | null;
  sessionId: string | null;
  launcherMsg: LauncherMsg;
  statusMessages: Array<string>;
};

type BooleanActionType =
  | 'leftSidebarVisible'
  | 'rightSidebarVisible'
  | 'rightSidebarAvailable';

export type Action =
  | {
      type: BooleanActionType;
      payload?: boolean;
    }
  | {
      type: 'windowIsFocused';
      payload: {isFocused: boolean; time: number};
    }
  | {
      type: 'LAUNCHER_MSG';
      payload: {
        severity: 'warning' | 'error';
        message: string;
      };
    }
  | {
      type: 'ADD_STATUS_MSG';
      payload: {msg: string; sender: string};
    }
  | {
      type: 'REMOVE_STATUS_MSG';
      payload: {msg: string; sender: string};
    };

export const initialState: () => State = () => ({
  leftSidebarVisible: true,
  rightSidebarVisible: true,
  rightSidebarAvailable: false,
  windowIsFocused: getRenderHostInstance().hasFocus(),
  activeSheet: null,
  share: null,
  sessionId: uuidv1(),
  launcherMsg: {
    severity: 'warning',
    message: '',
  },
  statusMessages: [],
  trackingTimeline: [],
});

function statusMessage(sender: string, msg: string): string {
  const messageTrimmed = msg.trim();
  const senderTrimmed = sender.trim();
  let statusMessage = senderTrimmed.length > 0 ? senderTrimmed : '';
  statusMessage =
    statusMessage.length > 0 && messageTrimmed.length > 0
      ? `${statusMessage}: ${messageTrimmed}`
      : '';
  return statusMessage;
}

export default function reducer(
  state: State | undefined,
  action: Actions,
): State {
  state = state || initialState();
  if (
    action.type === 'leftSidebarVisible' ||
    action.type === 'rightSidebarVisible' ||
    action.type === 'rightSidebarAvailable'
  ) {
    const newValue =
      typeof action.payload === 'undefined'
        ? !state[action.type]
        : action.payload;

    if (state[action.type] === newValue) {
      // value hasn't changed
      return state;
    } else {
      return {
        ...state,
        [action.type]: newValue,
      };
    }
  } else if (action.type === 'windowIsFocused') {
    return {
      ...state,
      windowIsFocused: action.payload.isFocused,
    };
  } else if (action.type === 'LAUNCHER_MSG') {
    return {
      ...state,
      launcherMsg: action.payload,
    };
  } else if (action.type === 'ADD_STATUS_MSG') {
    const {sender, msg} = action.payload;
    const statusMsg = statusMessage(sender, msg);
    if (statusMsg.length > 0) {
      return {
        ...state,
        statusMessages: [...state.statusMessages, statusMsg],
      };
    }
    return state;
  } else if (action.type === 'REMOVE_STATUS_MSG') {
    const {sender, msg} = action.payload;
    const statusMsg = statusMessage(sender, msg);
    if (statusMsg.length > 0) {
      const statusMessages = [...state.statusMessages];
      statusMessages.splice(statusMessages.indexOf(statusMsg), 1);
      return {...state, statusMessages};
    }
    return state;
  } else {
    return state;
  }
}

export const toggleAction = (
  type: BooleanActionType,
  payload?: boolean,
): Action => ({
  type,
  payload,
});

export const toggleLeftSidebarVisible = (payload?: boolean): Action => ({
  type: 'leftSidebarVisible',
  payload,
});

export const toggleRightSidebarVisible = (payload?: boolean): Action => ({
  type: 'rightSidebarVisible',
  payload,
});

export const toggleRightSidebarAvailable = (payload?: boolean): Action => ({
  type: 'rightSidebarAvailable',
  payload,
});

export const addStatusMessage = (payload: StatusMessageType): Action => ({
  type: 'ADD_STATUS_MSG',
  payload,
});

export const removeStatusMessage = (payload: StatusMessageType): Action => ({
  type: 'REMOVE_STATUS_MSG',
  payload,
});
