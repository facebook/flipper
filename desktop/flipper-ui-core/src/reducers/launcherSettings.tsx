/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {LauncherSettings} from 'flipper-common';
import {getRenderHostInstance} from 'flipper-frontend-core';
import {Actions} from './index';

export type Action = {
  type: 'UPDATE_LAUNCHER_SETTINGS';
  payload: LauncherSettings;
};

export default function reducer(
  state: LauncherSettings = getRenderHostInstance().serverConfig
    .launcherSettings,
  action: Actions,
): LauncherSettings {
  if (action.type === 'UPDATE_LAUNCHER_SETTINGS') {
    return action.payload;
  }
  return state;
}

export function updateLauncherSettings(settings: LauncherSettings): Action {
  return {
    type: 'UPDATE_LAUNCHER_SETTINGS',
    payload: settings,
  };
}
