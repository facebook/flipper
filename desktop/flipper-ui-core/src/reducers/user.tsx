/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {Actions} from './';
import {User} from 'flipper-common';

export type State = User;

export type Action =
  | {
      type: 'LOGIN';
      payload: User;
    }
  | {
      type: 'LOGOUT';
    };

const INITIAL_STATE: State = {};

export default function reducer(
  state: State = INITIAL_STATE,
  action: Actions,
): State {
  if (action.type === 'LOGOUT') {
    return {};
  } else if (action.type === 'LOGIN') {
    return {
      ...state,
      ...action.payload,
    };
  } else {
    return state;
  }
}

export const login = (payload: User): Action => ({
  type: 'LOGIN',
  payload,
});

export const logout = (): Action => ({
  type: 'LOGOUT',
});
