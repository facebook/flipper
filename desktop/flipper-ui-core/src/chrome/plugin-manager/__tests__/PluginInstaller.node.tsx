/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {default as PluginInstaller} from '../PluginInstaller';
import React from 'react';
import {render, waitFor} from '@testing-library/react';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import type {PluginDetails, UpdatablePluginDetails} from 'flipper-common';
import {Store} from '../../../reducers';
import {getRenderHostInstance} from 'flipper-frontend-core';

let getUpdatablePluginsMock: jest.Mock<any, any>;

beforeEach(() => {
  // flipperServer get resets before each test, no need to do so explicitly
  getUpdatablePluginsMock = getRenderHostInstance().flipperServer!.exec =
    jest.fn();
});

function getStore(installedPlugins: PluginDetails[] = []): Store {
  return configureStore([])({
    application: {sessionId: 'mysession'},
    plugins: {installedPlugins},
  }) as Store;
}

const samplePluginDetails1: UpdatablePluginDetails = {
  name: 'flipper-plugin-hello',
  entry: './test/index.js',
  version: '0.1.0',
  specVersion: 2,
  pluginType: 'client',
  main: 'dist/bundle.js',
  dir: '/Users/mock/.flipper/thirdparty/flipper-plugin-sample1',
  source: 'src/index.js',
  id: 'Hello',
  title: 'Hello',
  description: 'World?',
  isBundled: false,
  isActivatable: true,
  updateStatus: {
    kind: 'not-installed',
    version: '0.1.0',
  },
};

const samplePluginDetails2: UpdatablePluginDetails = {
  name: 'flipper-plugin-world',
  entry: './test/index.js',
  version: '0.2.0',
  specVersion: 2,
  pluginType: 'client',
  main: 'dist/bundle.js',
  dir: '/Users/mock/.flipper/thirdparty/flipper-plugin-sample2',
  source: 'src/index.js',
  id: 'World',
  title: 'World',
  description: 'Hello?',
  isBundled: false,
  isActivatable: true,
  updateStatus: {
    kind: 'not-installed',
    version: '0.2.0',
  },
};

const SEARCH_RESULTS = [samplePluginDetails1, samplePluginDetails2];

test('load PluginInstaller list', async () => {
  getUpdatablePluginsMock.mockReturnValue(Promise.resolve(SEARCH_RESULTS));
  const component = (
    <Provider store={getStore()}>
      <PluginInstaller
        // Bit ugly to have this as an effectively test-only option, but
        // without, we rely on height information from Electron which we don't
        // have, causing no items to be rendered.
        autoHeight
      />
    </Provider>
  );
  const {container, getByText} = render(component);
  await waitFor(() => getByText('hello'));
  expect(getUpdatablePluginsMock.mock.calls.length).toBe(1);
  expect(container).toMatchSnapshot();
});

test('load PluginInstaller list with one plugin installed', async () => {
  getUpdatablePluginsMock.mockReturnValue(
    Promise.resolve([
      {...samplePluginDetails1, updateStatus: {kind: 'up-to-date'}},
      samplePluginDetails2,
    ]),
  );
  const store = getStore([samplePluginDetails1]);
  const component = (
    <Provider store={store}>
      <PluginInstaller
        // Bit ugly to have this as an effectively test-only option, but
        // without, we rely on height information from Electron which we don't
        // have, causing no items to be rendered.
        autoHeight
      />
    </Provider>
  );
  const {container, getByText} = render(component);
  await waitFor(() => getByText('hello'));
  expect(getUpdatablePluginsMock.mock.calls.length).toBe(1);
  expect(container).toMatchSnapshot();
});
