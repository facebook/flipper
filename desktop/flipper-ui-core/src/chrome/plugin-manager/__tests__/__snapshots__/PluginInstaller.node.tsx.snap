// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load PluginInstaller list 1`] = `
<div>
  <div
    class="css-1v0y38i-Container e1k65efv0"
    height="500"
  >
    <div
      class="css-1lxv8hi-Container-Horizontal-SandyToolbarContainer e1ecpah20"
    >
      <span
        class="ant-input-group-wrapper ant-input-search"
      >
        <span
          class="ant-input-wrapper ant-input-group"
        >
          <input
            class="ant-input"
            placeholder="Search Flipper plugins..."
            type="text"
            value=""
          />
          <span
            class="ant-input-group-addon"
          >
            <button
              class="ant-btn ant-btn-default ant-btn-icon-only ant-input-search-button"
              type="button"
            >
              <span
                aria-label="search"
                class="anticon anticon-search"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="search"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                  />
                </svg>
              </span>
            </button>
          </span>
        </span>
      </span>
    </div>
    <div
      class="css-bgfc37-View-FlexBox-FlexColumn-Container emab7y20"
      tabindex="0"
    >
      <div
        class="ant-dropdown-trigger css-18abd42-View-FlexBox-FlexColumn e1e47qlf0"
      >
        <div
          class="css-1otvu18-View-FlexBox-FlexRow-TableHeadContainer eig1lcc1"
        >
          <div
            class="css-mpoiz3-TableHeadColumnContainer eig1lcc0"
            title="name"
            width="0"
          >
            <div
              class="eig1lcc3 css-x4q70f-InteractiveContainer-TableHeaderColumnInteractive e14xwmxq0"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1054ash-TableHeaderColumnContainer eig1lcc2"
              >
                Name
              </div>
            </div>
          </div>
          <div
            class="css-mpoiz3-TableHeadColumnContainer eig1lcc0"
            title="version"
            width="0"
          >
            <div
              class="eig1lcc3 css-x4q70f-InteractiveContainer-TableHeaderColumnInteractive e14xwmxq0"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1054ash-TableHeaderColumnContainer eig1lcc2"
              >
                Version
              </div>
            </div>
          </div>
          <div
            class="css-mpoiz3-TableHeadColumnContainer eig1lcc0"
            title="description"
            width="0"
          >
            <div
              class="eig1lcc3 css-x4q70f-InteractiveContainer-TableHeaderColumnInteractive e14xwmxq0"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1054ash-TableHeaderColumnContainer eig1lcc2"
              >
                Description
              </div>
            </div>
          </div>
          <div
            class="css-mpoiz3-TableHeadColumnContainer eig1lcc0"
            title="install"
            width="0"
          >
            <div
              class="eig1lcc3 css-x4q70f-InteractiveContainer-TableHeaderColumnInteractive e14xwmxq0"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1054ash-TableHeaderColumnContainer eig1lcc2"
              >
                Action
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-p5h61d-View-FlexBox-FlexColumn-Container emab7y20"
      >
        <div
          class="ant-dropdown-trigger css-18abd42-View-FlexBox-FlexColumn e1e47qlf0"
        >
          <div
            class="css-hg3ptm-View-FlexBox-FlexRow-TableBodyRowContainer e1pvjj0s1"
            data-key="flipper-plugin-hello"
          >
            <div
              class="css-yt4330-TableBodyColumnContainer e1pvjj0s0"
              title=""
              width="0"
            >
              <div
                style="display: inline-flex; align-items: center; overflow: hidden; gap: 8px;"
              >
                <span
                  class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
                  style="min-width: 0;"
                >
                  hello
                </span>
              </div>
            </div>
            <div
              class="css-pfp0fy-TableBodyColumnContainer e1pvjj0s0"
              title=""
              width="0"
            >
              <span
                class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
              >
                0.1.0
              </span>
            </div>
            <div
              class="css-yt4330-TableBodyColumnContainer e1pvjj0s0"
              title=""
              width="0"
            >
              <div
                class="css-s1wsbn-Container-Horizontal e1hsqii12"
              >
                <span
                  class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
                >
                  World?
                </span>
                <a
                  class="ant-typography"
                  href="https://yarnpkg.com/en/package/flipper-plugin-hello"
                >
                  <div
                    class="css-1kmzf9v-ColoredIconCustom ekc8qeh0"
                    color="var(--light-color-button-active)"
                    size="16"
                    src="https://facebook.com/assets/?name=info-circle&variant=filled&size=16&set=facebook_icons&density=1x"
                  />
                </a>
              </div>
            </div>
            <div
              class="css-16v1lq1-TableBodyColumnContainer e1pvjj0s0"
              title=""
              width="0"
            >
              <button
                class="ant-btn ant-btn-primary ant-btn-sm"
                type="button"
              >
                <span>
                  Install
                </span>
              </button>
            </div>
          </div>
          <div
            class="css-hg3ptm-View-FlexBox-FlexRow-TableBodyRowContainer e1pvjj0s1"
            data-key="flipper-plugin-world"
          >
            <div
              class="css-yt4330-TableBodyColumnContainer e1pvjj0s0"
              title=""
              width="0"
            >
              <div
                style="display: inline-flex; align-items: center; overflow: hidden; gap: 8px;"
              >
                <span
                  class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
                  style="min-width: 0;"
                >
                  world
                </span>
              </div>
            </div>
            <div
              class="css-pfp0fy-TableBodyColumnContainer e1pvjj0s0"
              title=""
              width="0"
            >
              <span
                class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
              >
                0.2.0
              </span>
            </div>
            <div
              class="css-yt4330-TableBodyColumnContainer e1pvjj0s0"
              title=""
              width="0"
            >
              <div
                class="css-s1wsbn-Container-Horizontal e1hsqii12"
              >
                <span
                  class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
                >
                  Hello?
                </span>
                <a
                  class="ant-typography"
                  href="https://yarnpkg.com/en/package/flipper-plugin-world"
                >
                  <div
                    class="css-1kmzf9v-ColoredIconCustom ekc8qeh0"
                    color="var(--light-color-button-active)"
                    size="16"
                    src="https://facebook.com/assets/?name=info-circle&variant=filled&size=16&set=facebook_icons&density=1x"
                  />
                </a>
              </div>
            </div>
            <div
              class="css-16v1lq1-TableBodyColumnContainer e1pvjj0s0"
              title=""
              width="0"
            >
              <button
                class="ant-btn ant-btn-primary ant-btn-sm"
                type="button"
              >
                <span>
                  Install
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-1lxv8hi-Container-Horizontal-SandyToolbarContainer e1ecpah20"
    >
      <div
        class="ant-row ant-row-no-wrap ant-row-middle"
        style="margin-left: -4px; margin-right: -4px;"
      >
        <div
          class="ant-col"
          style="padding-left: 4px; padding-right: 4px; flex-basis: auto; min-width: 0;"
        >
          <input
            class="ant-input"
            placeholder="Select a Flipper package or just drag and drop it here..."
            type="text"
            value=""
          />
        </div>
        <div
          class="ant-col"
          style="padding-left: 4px; padding-right: 4px; flex: 0 0 auto; min-width: 0;"
        >
          <button
            class="ant-btn ant-btn-default ant-btn-icon-only"
            disabled=""
            title="Reset"
            type="button"
          >
            <span
              aria-label="close"
              class="anticon anticon-close"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="close"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="ant-col"
          style="padding-left: 4px; padding-right: 4px; flex: 0 0 auto; min-width: 0;"
        >
          <button
            class="ant-btn ant-btn-default ant-btn-icon-only"
            title="Select a Flipper package or just drag and drop it here..."
            type="button"
          >
            <span
              aria-label="upload"
              class="anticon anticon-upload"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="upload"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
      <div
        class="css-5ukfaz-View-FlexBox-FlexRow-ButtonContainer eguixfz1"
      >
        <div
          class="css-wospjg-View-FlexBox-FlexRow ek54xq0"
        >
          <button
            class="ant-btn ant-btn-primary"
            disabled=""
            type="button"
          >
            <span>
              Install
            </span>
          </button>
          <div
            class="css-170i4ha-View-FlexBox-FlexRow-ErrorGlyphContainer eguixfz0"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`load PluginInstaller list with one plugin installed 1`] = `
<div>
  <div
    class="css-1v0y38i-Container e1k65efv0"
    height="500"
  >
    <div
      class="css-1lxv8hi-Container-Horizontal-SandyToolbarContainer e1ecpah20"
    >
      <span
        class="ant-input-group-wrapper ant-input-search"
      >
        <span
          class="ant-input-wrapper ant-input-group"
        >
          <input
            class="ant-input"
            placeholder="Search Flipper plugins..."
            type="text"
            value=""
          />
          <span
            class="ant-input-group-addon"
          >
            <button
              class="ant-btn ant-btn-default ant-btn-icon-only ant-input-search-button"
              type="button"
            >
              <span
                aria-label="search"
                class="anticon anticon-search"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="search"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                  />
                </svg>
              </span>
            </button>
          </span>
        </span>
      </span>
    </div>
    <div
      class="css-bgfc37-View-FlexBox-FlexColumn-Container emab7y20"
      tabindex="0"
    >
      <div
        class="ant-dropdown-trigger css-18abd42-View-FlexBox-FlexColumn e1e47qlf0"
      >
        <div
          class="css-1otvu18-View-FlexBox-FlexRow-TableHeadContainer eig1lcc1"
        >
          <div
            class="css-mpoiz3-TableHeadColumnContainer eig1lcc0"
            title="name"
            width="0"
          >
            <div
              class="eig1lcc3 css-x4q70f-InteractiveContainer-TableHeaderColumnInteractive e14xwmxq0"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1054ash-TableHeaderColumnContainer eig1lcc2"
              >
                Name
              </div>
            </div>
          </div>
          <div
            class="css-mpoiz3-TableHeadColumnContainer eig1lcc0"
            title="version"
            width="0"
          >
            <div
              class="eig1lcc3 css-x4q70f-InteractiveContainer-TableHeaderColumnInteractive e14xwmxq0"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1054ash-TableHeaderColumnContainer eig1lcc2"
              >
                Version
              </div>
            </div>
          </div>
          <div
            class="css-mpoiz3-TableHeadColumnContainer eig1lcc0"
            title="description"
            width="0"
          >
            <div
              class="eig1lcc3 css-x4q70f-InteractiveContainer-TableHeaderColumnInteractive e14xwmxq0"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1054ash-TableHeaderColumnContainer eig1lcc2"
              >
                Description
              </div>
            </div>
          </div>
          <div
            class="css-mpoiz3-TableHeadColumnContainer eig1lcc0"
            title="install"
            width="0"
          >
            <div
              class="eig1lcc3 css-x4q70f-InteractiveContainer-TableHeaderColumnInteractive e14xwmxq0"
              style="z-index: auto; right: 0px; bottom: 0px; width: 100%; height: 100%;"
            >
              <div
                class="css-1054ash-TableHeaderColumnContainer eig1lcc2"
              >
                Action
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-p5h61d-View-FlexBox-FlexColumn-Container emab7y20"
      >
        <div
          class="ant-dropdown-trigger css-18abd42-View-FlexBox-FlexColumn e1e47qlf0"
        >
          <div
            class="css-hg3ptm-View-FlexBox-FlexRow-TableBodyRowContainer e1pvjj0s1"
            data-key="flipper-plugin-hello"
          >
            <div
              class="css-yt4330-TableBodyColumnContainer e1pvjj0s0"
              title=""
              width="0"
            >
              <div
                style="display: inline-flex; align-items: center; overflow: hidden; gap: 8px;"
              >
                <span
                  class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
                  style="min-width: 0;"
                >
                  hello
                </span>
              </div>
            </div>
            <div
              class="css-pfp0fy-TableBodyColumnContainer e1pvjj0s0"
              title=""
              width="0"
            >
              <span
                class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
              >
                0.1.0
              </span>
            </div>
            <div
              class="css-yt4330-TableBodyColumnContainer e1pvjj0s0"
              title=""
              width="0"
            >
              <div
                class="css-s1wsbn-Container-Horizontal e1hsqii12"
              >
                <span
                  class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
                >
                  World?
                </span>
                <a
                  class="ant-typography"
                  href="https://yarnpkg.com/en/package/flipper-plugin-hello"
                >
                  <div
                    class="css-1kmzf9v-ColoredIconCustom ekc8qeh0"
                    color="var(--light-color-button-active)"
                    size="16"
                    src="https://facebook.com/assets/?name=info-circle&variant=filled&size=16&set=facebook_icons&density=1x"
                  />
                </a>
              </div>
            </div>
            <div
              class="css-16v1lq1-TableBodyColumnContainer e1pvjj0s0"
              title=""
              width="0"
            >
              <button
                class="ant-btn ant-btn-default ant-btn-sm"
                type="button"
              >
                <span>
                  Remove
                </span>
              </button>
            </div>
          </div>
          <div
            class="css-hg3ptm-View-FlexBox-FlexRow-TableBodyRowContainer e1pvjj0s1"
            data-key="flipper-plugin-world"
          >
            <div
              class="css-yt4330-TableBodyColumnContainer e1pvjj0s0"
              title=""
              width="0"
            >
              <div
                style="display: inline-flex; align-items: center; overflow: hidden; gap: 8px;"
              >
                <span
                  class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
                  style="min-width: 0;"
                >
                  world
                </span>
              </div>
            </div>
            <div
              class="css-pfp0fy-TableBodyColumnContainer e1pvjj0s0"
              title=""
              width="0"
            >
              <span
                class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
              >
                0.2.0
              </span>
            </div>
            <div
              class="css-yt4330-TableBodyColumnContainer e1pvjj0s0"
              title=""
              width="0"
            >
              <div
                class="css-s1wsbn-Container-Horizontal e1hsqii12"
              >
                <span
                  class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
                >
                  Hello?
                </span>
                <a
                  class="ant-typography"
                  href="https://yarnpkg.com/en/package/flipper-plugin-world"
                >
                  <div
                    class="css-1kmzf9v-ColoredIconCustom ekc8qeh0"
                    color="var(--light-color-button-active)"
                    size="16"
                    src="https://facebook.com/assets/?name=info-circle&variant=filled&size=16&set=facebook_icons&density=1x"
                  />
                </a>
              </div>
            </div>
            <div
              class="css-16v1lq1-TableBodyColumnContainer e1pvjj0s0"
              title=""
              width="0"
            >
              <button
                class="ant-btn ant-btn-primary ant-btn-sm"
                type="button"
              >
                <span>
                  Install
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-1lxv8hi-Container-Horizontal-SandyToolbarContainer e1ecpah20"
    >
      <div
        class="ant-row ant-row-no-wrap ant-row-middle"
        style="margin-left: -4px; margin-right: -4px;"
      >
        <div
          class="ant-col"
          style="padding-left: 4px; padding-right: 4px; flex-basis: auto; min-width: 0;"
        >
          <input
            class="ant-input"
            placeholder="Select a Flipper package or just drag and drop it here..."
            type="text"
            value=""
          />
        </div>
        <div
          class="ant-col"
          style="padding-left: 4px; padding-right: 4px; flex: 0 0 auto; min-width: 0;"
        >
          <button
            class="ant-btn ant-btn-default ant-btn-icon-only"
            disabled=""
            title="Reset"
            type="button"
          >
            <span
              aria-label="close"
              class="anticon anticon-close"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="close"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="ant-col"
          style="padding-left: 4px; padding-right: 4px; flex: 0 0 auto; min-width: 0;"
        >
          <button
            class="ant-btn ant-btn-default ant-btn-icon-only"
            title="Select a Flipper package or just drag and drop it here..."
            type="button"
          >
            <span
              aria-label="upload"
              class="anticon anticon-upload"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="upload"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M400 317.7h73.9V656c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V317.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 163a8 8 0 00-12.6 0l-112 141.7c-4.1 5.3-.4 13 6.3 13zM878 626h-60c-4.4 0-8 3.6-8 8v154H214V634c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v198c0 17.7 14.3 32 32 32h684c17.7 0 32-14.3 32-32V634c0-4.4-3.6-8-8-8z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
      <div
        class="css-5ukfaz-View-FlexBox-FlexRow-ButtonContainer eguixfz1"
      >
        <div
          class="css-wospjg-View-FlexBox-FlexRow ek54xq0"
        >
          <button
            class="ant-btn ant-btn-primary"
            disabled=""
            type="button"
          >
            <span>
              Install
            </span>
          </button>
          <div
            class="css-170i4ha-View-FlexBox-FlexRow-ErrorGlyphContainer eguixfz0"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
